log4j-api/src/main/java/org/apache/logging/log4j/spi/Provider.java

```diff
@@ -97,127 +217,266 @@ public String getVersions() {
     }
 
     /**
-     * Gets the class name of the {@link org.apache.logging.log4j.spi.LoggerContextFactory} implementation of this
-     * Provider.
+     * Gets the class name of the {@link LoggerContextFactory} implementation of this Provider.
      *
-     * @return the class name of a LoggerContextFactory implementation
+     * @return the class name of a LoggerContextFactory implementation or {@code null} if unspecified.
+     * @see #loadLoggerContextFactory()
      */
     public String getClassName() {
-        if (loggerContextFactoryClass != null) {
-            return loggerContextFactoryClass.getName();
-        }
-        return className;
+        return loggerContextFactoryClass != null ? loggerContextFactoryClass.getName() : className;
     }
 
     /**
-     * Loads the {@link org.apache.logging.log4j.spi.LoggerContextFactory} class specified by this Provider.
+     * Loads the {@link LoggerContextFactory} class specified by this Provider.
      *
-     * @return the LoggerContextFactory implementation class or {@code null} if there was an error loading it
+     * @return the LoggerContextFactory implementation class or {@code null} if unspecified or a loader error occurred.
+     * @see #createLoggerContextFactory()
      */
     public Class<? extends LoggerContextFactory> loadLoggerContextFactory() {
         if (loggerContextFactoryClass != null) {
             return loggerContextFactoryClass;
         }
-        if (className == null) {
-            return null;
-        }
+        final String className = getClassName();
         final ClassLoader loader = classLoader.get();
-        if (loader == null) {
+        // Support for deprecated {@code META-INF/log4j-provider.properties} format.
+        // In the remaining cases {@code loader == null}.
+        if (loader == null || className == null) {
             return null;
         }
         try {
             final Class<?> clazz = loader.loadClass(className);
             if (LoggerContextFactory.class.isAssignableFrom(clazz)) {
                 return clazz.asSubclass(LoggerContextFactory.class);
+            } else {
+                LOGGER.error(
+                        "Class {} specified in {} does not extend {}",
+                        className,
+                        getUrl(),
+                        LoggerContextFactory.class.getName());
             }
         } catch (final Exception e) {
-            LOGGER.error("Unable to create class {} specified in {}", className, url.toString(), e);
+            LOGGER.error("Unable to create class {} specified in {}", className, getUrl(), e);
         }
         return null;
     }
 
     /**
-     * Gets the class name of the {@link org.apache.logging.log4j.spi.ThreadContextMap} implementation of this Provider.
-     *
+     * Extension point for providers to create a {@link LoggerContextFactory}.
+     * @since 2.24.0
+     */
+    protected LoggerContextFactory createLoggerContextFactory() {
```

Comment:
I have a feeling this is here only for tests. I cannot see it getting it used in a way that necessitates a `public` visibility. Am I missing something?

Reply:
This is actually one of the two methods that implementations should implement.

Due to the long history of the class, there are other methods:

- `#getClassName()` gives the name of the class, but then Log4j API must guess the classloader to use to load it,
- `#loadLoggerContextFactory()` returns a class, but unless I am mistaken JPMS will not allow Log4j API to instantiate the class unless it is `public` and in an exported package,

I introduced this method so that implementations can instantiate it in any way they need. E.g. Log4j Core 3.x uses:
```java
new Log4jContextFactory(instanceFactory);
```